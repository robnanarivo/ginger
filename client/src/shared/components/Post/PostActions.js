import React from 'react';
import CardActions from '@mui/material/CardActions';
import VisibilityOffIcon from '@mui/icons-material/VisibilityOff';
import DeleteIcon from '@mui/icons-material/Delete';
import FlagIcon from '@mui/icons-material/Flag';
import Button from '@mui/material/Button';
import { Alert } from '@mui/material';
import { useHistory, useParams } from 'react-router-dom';
import { getProfile, updateProfile } from '../../../api';

const PostActions = (props) => {
  const {
    postId,
    handleDelete,
    creatorId,
    admins,
  } = props;
  const history = useHistory();
  const { groupId } = useParams();
  const userId = sessionStorage.getItem('userId');
  const token = sessionStorage.getItem('userToken');
  const [flagged, setFlagged] = React.useState(false);
  const [error, setError] = React.useState('');
  // console.log(creatorId, userId, admins);

  const handleFlag = async () => {
    if (flagged) return;
    const profile = await getProfile(userId, token);
    if ('error' in profile) {
      setError(profile.error);
      return;
    }
    const newFlags = [...profile.flags, postId];
    const updatedProfile = await updateProfile(userId, { flags: newFlags }, token);
    if ('error' in updatedProfile) {
      setError(updatedProfile.error);
      return;
    }
    setFlagged(true);
  };

  const handleHide = async () => {
    const profile = await getProfile(userId, token);
    if ('error' in profile) {
      setError(profile.error);
      return;
    }
    const newHides = [...profile.hides, postId];
    const updatedProfile = await updateProfile(userId, { hides: newHides }, token);
    if ('error' in updatedProfile) {
      setError(updatedProfile.error);
      return;
    }
    history.push(`/groupboard/${groupId}`);
  };

  React.useEffect(async () => {
    const profile = await getProfile(userId, token);
    if ('error' in profile) {
      setError(profile.error);
      return;
    }
    if (profile.flags.includes(postId)) {
      setFlagged(true);
    }
  }, []);

  return (
    <CardActions
      disableSpacing
      sx={{
        display: 'flex',
        marginLeft: '6px',
      }}
    >
      <Button
        sx={{ color: '#A9A9A9' }}
        startIcon={<VisibilityOffIcon />}
        style={{ textTransform: 'none' }}
        onClick={handleHide}
      >
        Hide
      </Button>
      <Button
        sx={{ color: flagged ? '#ffa726' : '#A9A9A9' }}
        startIcon={<FlagIcon />}
        style={{ textTransform: 'none' }}
        onClick={handleFlag}
      >
        Flag
      </Button>
      {
        creatorId === userId || admins.includes(userId)
          ? (
            <Button
              sx={{ color: '#A9A9A9' }}
              startIcon={<DeleteIcon />}
              style={{ textTransform: 'none' }}
              onClick={handleDelete}
            >
              Delete
            </Button>
          ) : (
            ''
          )
      }

      { error && <Alert severity="error" variant="outlined">{error}</Alert> }
    </CardActions>
  );
};

export default PostActions;
