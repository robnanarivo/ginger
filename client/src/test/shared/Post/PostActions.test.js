import React from 'react';
import renderer from 'react-test-renderer';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import PostActions from '../../../shared/components/Post/PostActions';
import 'jest-styled-components';

describe('PostActions unit test', () => {
  it('PostActions Snapshot test', () => {
    const history = createMemoryHistory();
    const tree = renderer
      .create(
        <Router history={history}>
          <PostActions creatorName="Bob" admins={[]} />
        </Router>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
