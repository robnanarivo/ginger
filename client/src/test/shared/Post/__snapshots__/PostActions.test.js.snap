// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostActions unit test PostActions Snapshot test 1`] = `
<div
  className="MuiCardActions-root css-k3c8wq-MuiCardActions-root"
>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-rn8hnj-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "textTransform": "none",
      }
    }
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="VisibilityOffIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
        />
      </svg>
    </span>
    Hide
  </button>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-rn8hnj-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "textTransform": "none",
      }
    }
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="FlagIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M14.4 6 14 4H5v17h2v-7h5.6l.4 2h7V6z"
        />
      </svg>
    </span>
    Flag
  </button>
  
  
</div>
`;
